flax/nnx/transforms/iteration.py

```diff
@@ -1373,6 +1384,9 @@ def scan_wrapper(*args, **kwargs):
       pure_out,
     ) = scan_out
 
+    if pure_carry_arg_out is not None and scan_fn.carry_graphdef is not None:
+      pure_carry_arg_out = extract.NodeStates.from_split(scan_fn.carry_graphdef, pure_carry_arg_out)
```

Comment:
Finally, in `scan_wrapper`, we reconstruct the carry as NodeStates into `pure_carry_arg_out`.